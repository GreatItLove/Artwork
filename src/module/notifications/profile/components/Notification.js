import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import _map from 'lodash/map';
import _head from 'lodash/head';
import _toUpper from 'lodash/toUpper';
import TimeAgo from 'react-timeago';
import { RefreshIndicator, Avatar } from 'material-ui';
import ActionReportProblem from 'material-ui/svg-icons/action/report-problem';

const propTypes = {
  fetchLogs: PropTypes.func.isRequired,
  logsLoading: PropTypes.bool.isRequired,
  logsData: PropTypes.array,
  logsError: PropTypes.string
};

class Notification extends Component {
  constructor() {
    super();
    this.state = {
      toggle: false
    };
  }

  showAllNotifications() {
    const { logsLoading, logsData, logsError } = this.props;
    return logsLoading ? (
      <div>
        <RefreshIndicator left={150} top={100} status="loading" />
      </div>
    ) : logsData.length ? (
      <div>
        {_map(logsData, (notification, key) => {
          return (
            <div key={key} className="notification-single">
              <div className="avatar">
                <Avatar
                  size={30}
                  icon={!notification.fullName ? <ActionReportProblem /> : null}
                  backgroundColor={this.statusColor(notification.logType)}>
                  {notification.fullName
                    ? _toUpper(_head(notification.fullName))
                    : ''}
                </Avatar>
              </div>
              <p className="details">
                {notification.name ? <a href="">{notification.name}</a> : ''}
                <span>{this.activityMsg(notification)}</span>
              </p>
              <p className="time">
                <TimeAgo date={parseInt(notification.dateCreated, 10) * 1000} />
              </p>
            </div>
          );
        })}
      </div>
    ) : (
          <div>{logsError}</div>
        );
  }
  activityMsg = notification => {
    const { fullName, resourceType, details, logType } = notification;
    let fields, id;
    if (resourceType !== 'login') {
      id = JSON.parse(details).id;
      fields = JSON.parse(details).fields;
    }
    if (resourceType === 'login') return details;
    if (logType === 'edit') return `${fullName} updated ${resourceType} ${id}. Updated field(s): ${fields}`;
    if (logType === 'add') return `${fullName} created a new ${resourceType} - ${id}`;
    if (logType === 'delete') return `${fullName} deleted a ${resourceType} - ${id}`;
  };
  statusColor = type => {
    if (type === 'add' || type === 'Successful Login') return '#98e099';
    if (type === 'edit') return '#69a6de';
    if (type === 'delete' || type === 'Failed Login') return '#ec7e81';
  };
  showNotification = () => {
    const { toggle } = this.state;
    if (!toggle) {
      this.props.fetchLogs({ limit: 20 });
    }
    this.setState({ toggle: !toggle });
  };
  render() {
    const toggleNotification = classNames({
      'slide-content': true,
      toggle: this.state.toggle
    });
    return (
      <div>
        <button className="notification-trigger an-circle-icon-btn notification" onClick={this.showNotification}>
          <i className="fa fa-flash" />
        </button>
        <div className={toggleNotification}>
          <h6>Latest Activities</h6>
          {this.showAllNotifications()}
        </div>
      </div>
    );
  }
}

Notification.propTypes = propTypes;

export default Notification;
