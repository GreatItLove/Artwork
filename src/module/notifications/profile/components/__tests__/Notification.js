import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';

import Notification from '../Notification';

describe('Notification', () => {
  const props = {
    logsLoading: true,
    logsData: [],
    logsError: '',
    fetchLogs: jest.fn(),
  };
  const logsData = [
    {
      browserInfo: `Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/61.0.3163.100 Safari/537.36`,
      customerId: `1859`,
      dateCreated: `1507019251`,
      details: `The user logged-in successfully.`,
      id: 366,
      ip: `127.0.0.1`,
      logType: `Successful Login`,
      resourceType: `login`,
      userId: `6733`,
    },
    {
      browserInfo: `Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/61.0.3163.100 Safari/537.36`,
      customerId: `1859`,
      dateCreated: `1507017081`,
      details: `The user logged-in successfully.`,
      id: 365,
      ip: `127.0.0.1`,
      logType: `Successful Login`,
      resourceType: `login`,
      userId: `6757`,
    },
  ];

  it('Snapshot of Notification when logsLoading is true', () => {
    const wrapper = shallow(<Notification {...props} />);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
  it('Snapshot of Notification when logsLoading is false ', () => {
    const propsLoadingFalse = {
      ...props,
      ...{
        logsLoading: false,
        logsData: logsData,
      },
    };
    const wrapper = shallow(<Notification {...propsLoadingFalse} />);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
  it('Snapshot of Notification when error ', () => {
    const propsLogsError = {
      ...props,
      ...{
        logsLoading: false,
        logsError: 'err',
      },
    };
    const wrapper = shallow(<Notification {...propsLogsError} />);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});
