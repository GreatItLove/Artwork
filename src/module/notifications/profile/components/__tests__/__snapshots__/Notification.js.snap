// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification Snapshot of Notification when error  1`] = `
<div>
  <button
    className="notification-trigger an-circle-icon-btn notification"
    onClick={[Function]}
  >
    <i
      className="fa fa-flash"
    />
  </button>
  <div
    className="slide-content"
  >
    <h6>
      Latest Activities
    </h6>
    <div>
      err
    </div>
  </div>
</div>
`;

exports[`Notification Snapshot of Notification when logsLoading is false  1`] = `
<div>
  <button
    className="notification-trigger an-circle-icon-btn notification"
    onClick={[Function]}
  >
    <i
      className="fa fa-flash"
    />
  </button>
  <div
    className="slide-content"
  >
    <h6>
      Latest Activities
    </h6>
    <div>
      <div
        className="notification-single"
      >
        <p
          className="details"
        >
          <span>
            The user logged-in successfully.
          </span>
        </p>
        <p
          className="time"
        >
          <TimeAgo
            component="time"
            date={1507019251000}
            formatter={[Function]}
            live={true}
            maxPeriod={Infinity}
            minPeriod={0}
            now={[Function]}
          />
        </p>
      </div>
      <div
        className="notification-single"
      >
        <p
          className="details"
        >
          <span>
            The user logged-in successfully.
          </span>
        </p>
        <p
          className="time"
        >
          <TimeAgo
            component="time"
            date={1507017081000}
            formatter={[Function]}
            live={true}
            maxPeriod={Infinity}
            minPeriod={0}
            now={[Function]}
          />
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`Notification Snapshot of Notification when logsLoading is true 1`] = `
<div>
  <button
    className="notification-trigger an-circle-icon-btn notification"
    onClick={[Function]}
  >
    <i
      className="fa fa-flash"
    />
  </button>
  <div
    className="slide-content"
  >
    <h6>
      Latest Activities
    </h6>
    <div
      style={
        Object {
          "left": "45%",
          "position": "absolute",
          "top": "30%",
        }
      }
    >
      <RefreshIndicator
        left={0}
        percentage={0}
        size={40}
        status="loading"
        top={0}
      />
    </div>
  </div>
</div>
`;
